dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+    Cb&& cb) {
+  if (sample_desc.shape.lanes_axis == 2) {
+    cb(create_direct_conv<2>(sample_desc, in_loader, block_setup));
+  } else {
+    assert(sample_desc.shape.lanes_axis == 1);
+    cb(create_direct_conv<1>(sample_desc, in_loader, block_setup));
+  }
+}
+
+/*
+Given a HWC image and RS filter, all the necessary products for computing the convolution
+explicitly can be seen as multiplying a matrix of shape HWC x RS with a vector of size RS. Now,
+assuming a standard contiguious memory layout of the HWC image, if you look at any given column of
+the HWC x RS matrix, consecutive rows map to consecutive memory addresses (with the exception of
+positions when we cross H, W extents and border remapping takes place). This implementation melds
+the W and C extents to utilize this property. Additionally, the implementation that uses shared
+memory loads the inputs in blocks with extents corresponding to (D, )H, W * C extents to account
+for the fact that spatailly close products will reuse some of the inputs.
+*/
+template <typename SampleDescT, typename BlockSetupFactory, typename InLoaderFactory,
+          typename GridSetupT>
+__global__ void filter(const SampleDescT* __restrict__ descs, BlockSetupFactory block_setup_factory,
+                       InLoaderFactory in_loader_factory, GridSetupT grid_setup) {
+  extern __shared__ char shm[];
+  int sample_idx = grid_setup.sample_idx();
+  for (int sample_idx = grid_setup.sample_idx(); sample_idx < grid_setup.num_samples();
+       sample_idx += grid_setup.sample_dim()) {
+    auto sample_desc = descs[sample_idx];
+    const auto& block_setup = block_setup_factory(sample_idx);
+    const auto& log_block_extents = sample_desc.shape.log_block_extents;
+    auto block_start = grid_setup.block_idx() * log_block_extents;
+    auto process_sample = [&](const auto& out_extents, const auto& out_strides,
+                              const auto& out_frame_stride) {
+      if (any_coord(block_start >= out_extents)) {
+        return;  // early exit to avoid all the setup only to do nothing in the stride loop
+      }
+      auto grid_size = grid_setup.grid_dim() * log_block_extents;
+      const auto& in_loader = in_loader_factory(sample_idx);
+      if (sample_desc.shape.in_workspace_extents.x) {
+        using In = typename SampleDescT::In;
+        int* idx_workspace = reinterpret_cast<int*>(shm);
+        In* in_workspace = reinterpret_cast<In*>(shm + sample_desc.shape.in_workspace_offset);
+        with_shm_conv(
+            sample_desc, in_loader, block_setup, idx_workspace, in_workspace, [&](auto&& conv) {
+              stride_grid(block_start, grid_size, out_extents, out_strides, out_frame_stride, conv);
+            });
+
+      } else {
+        with_direct_conv(sample_desc, in_loader, block_setup, [&](auto&& conv) {
+          stride_grid(block_start, grid_size, out_extents, out_strides, out_frame_stride, conv);
+        });
```

Comment:
Based on our experience with optimizing stuff for H100, I would recommend splitting this kernel into two variants (and possibly launching both). The use of SHM will prevent the shared memory from being used as L1 cache, which will cripple the direct variant. There should be two launches - the samples eligible for shm variant should go to shm kernel and the others should go to the direct kernel, with reduced SHM consumption.

Reply:
done